ports/stm32/ulpi.c

```diff
@@ -0,0 +1,179 @@
+  * @param  Addr the Address of the ULPI Register
+  * @param  Data Data to write
+  * @retval Returns value of PHY CR register
+  */
+static uint32_t USB_ULPI_Write(uint32_t Addr, uint32_t Data) {
+    __IO uint32_t val;
+    __IO uint32_t timeout = 10;   /* Can be tuned based on the Clock or etc... */
+
+    USB_OTG_WRITE_REG32(USBULPI_PHYCR, USBULPI_New | USBULPI_RW | (Addr << 16) | (Data & 0x000000ff));
+    val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    while (((val & USBULPI_S_DONE) == 0) && (timeout--)) {
+        val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    }
+
+    val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    return 0;
+}
+
+/**
+  * @brief  This function configures the USB PHY to enter the low power mode
+  * @param  None
+  * @retval None
+  */
+void ulpi_enter_low_power(void) {
+    __IO uint32_t regval = 0;
+
+    /* disable ULPI_CLK by accessing ULPI_PHY */
+    /* read Vendor ID : (Low, High) 0x24,0x04 for USB3300 */
+    regval = USB_ULPI_Read(0x00);
+    if (regval != 0x24) {
+        __fatal_error("ULPI Error 0x00 != 0x24");
+    }
+
+    regval = USB_ULPI_Read(0x01);
+    if (regval != 0x04) {
+        __fatal_error("ULPI Error 0x01 != 0x04");
+    }
+
+    /* read Product ID */
+    regval = USB_ULPI_Read(0x02);
+    if (regval != 0x07) {
+        __fatal_error("ULPI Error 0x02 != 0x07");
+    }
+
+    regval = USB_ULPI_Read(0x03);
+    if (regval != 0x00) {
+        __fatal_error("ULPI Error 0x03 != 0x00");
+    }
+
+    /* Write to scratch register the pattern 0x55 */
+    USB_ULPI_Write(0x16, 0x55);
+    /* Read to scratch Register and check-it again the written Pattern */
+    regval = USB_ULPI_Read(0x16);
+    if (regval != 0x55) {
+        __fatal_error("ULPI Error 0x16 != 0x55");
+    }
+
+    /* Write to scratch register the pattern 0xAA */
+    USB_ULPI_Write(0x16, 0xAA);
+    /* Read to scratch Register and check-it again the written Pattern */
+    regval = USB_ULPI_Read(0x16);
+    if (regval != 0xAA) {
+        __fatal_error("ULPI Error 0x16 != 0xAA");
+    }
+
+    /* read InterfaceControl reg */
+    regval = USB_ULPI_Read(0x07);
+    /* write InterfaceControl reg,to disable PullUp on stp,
+       to avoid USB_PHY wake up when MCU entering standby */
+    USB_ULPI_Write(0x07, regval | 0x80);
+    /* read InterfaceControl reg */
+    regval = USB_ULPI_Read(0x07);
+    if (regval != 0x80) {
+        __fatal_error("ULPI Error 0x07 != 0x80");
+    }
+
+    /* read FunctionControl reg */
+    regval = USB_ULPI_Read(0x04);
+
+    // Reg 0x40 has a different value if USB is disconnected.
+    if (regval != 0x40 && regval != 0x45) {
+        __fatal_error("ULPI Error 0x04 != 0x40 || 0x45");
+    }
+
+    /* write FunctionControl reg,to put PHY into LowPower mode */
+    USB_ULPI_Write(0x04, regval & (~0x40));
+
+    /* read FunctionControl reg again */
+    regval = USB_ULPI_Read(0x04);
+    if (regval != 0x00) {
+        __fatal_error("ULPI Error 0x04 != 0x00");
+    }
+
+    /* Delay 4 ms */
+    HAL_Delay(10);
```

Comment:
Are SYSTICK IRQs guaranteed to be running when this function is called?

Also, the comment doesn't match the code, 4 vs 10.

Reply:
Doesn't `HAL_Delay` fallback to `mp_hal_delay_ms` if IRQs are disabled ?

Reply:
Yes you are right, so it should be OK.

Reply:
Note this function may need to be called from `EARLY_INIT` or even `STARTUP` if it was used on standby entry, ULPI may still be in low-power mode, so it's best to call `HAL_Delay` first to avoid running the event loop before everything is initialized. On that matter, should there be a check that MP is initialized  in `mp_hal_delay_ms` somehow before calling the event loop for cases like these ?

Reply:
> On that matter, should there be a check that MP is initialized in mp_hal_delay_ms somehow before calling the event loop for cases like these ?

I don't think it needs a check.  If you call `HAL_Delay()` then it will only fall back to `mp_hal_delay_ms()` in the case that IRQs are disabled, in which case it goes to the "else" part of the if in that function, which does not call `MICROPY_EVENT_POLL_HOOK`.

Summary: calling `HAL_Delay()` here should be OK.  But please make the comment match the delay used, and put in a comment about why this delay is needed.

Reply:
The comment matches now. The small delay because CLKOUT keeps running after the SUSPENDM bit is cleared for a while before stopping, so it's probably safer to wait to ensure ULPI is in low-power mode before returning from this function.